### PR TITLE
flux-jobs: take optional job ids as position arguments

### PR DESCRIPTION
**Pull request by chu11**
Per issue #2868.  Just thought I'd throw this up for comment as the implementation is ... meh

- Binding for `job-info.list-id` was not available, so had to add one.

- binding (just like with `list` and `list-inactive`) returns an RPC object that user can get job returned via `get_job()`.

The above is nothing special, but AFAICT it leads to the following problem:

- b/c the binding doesn't return a future, the normal mechanisms to run things asynchronously (i.e. with `then` callbacks) is not available (or atleast I can't see a way).  So the retrieval of each jobid's job information is synchronous.  Perhaps not horrible for the general case when a user just wants a few job IDs listed, but eventually it'll be a problem when someone scripts  `flux-jobs` with a ton of IDs.
  - OR do we expect our power users to script against our APIs and not against `flux-jobs`?

Possible solutions:

- change `job-info.list-id` service to take a list of job ids and returns an array of jobs, so it's a lot more like `list` and `list-inactive`.  Since the initial need in issue #2592 just needed a single ID, that's all that was supported.

- We can "manually" build `job-info.list-id` RPCs in `flux-jobs` or in some new bindings that will deal with the futures / and fetching the job ids asynchronously.

But before going down either path wanted to see what thoughts were.  Is the cheapo-synchronous way more than sufficient for now?

### CODE REVIEW

**Comment by grondo** [comment]
> b/c the binding doesn't return a future, the normal mechanisms to run things asynchronously (i.e. with then callbacks) is not available (or atleast I can't see a way).

I'm probably not understanding the issue, but `RPC` is a subclass of `Future` and it is fine to call `.then` method on it...

**Comment by chu11** [comment]
> I'm probably not understanding the issue, but RPC is a subclass of Future and it is fine to call .then method on it...

Ahhh, I had missed that :-)

**Comment by chu11** [comment]
re-pushed, now using then / reactor callback to get the jobs.  woot!  Thanks for the now obvious suggestion @grondo  :-)

Not entirely sure if how I implemented the callback is really Pythonic.

```
+def list_id_cb(future, arg):
+    (flux_handle, rpc_handle, jobid, count, args, jobs) = arg
```

Basically pass in a tuple of everything I need in the callback.  Basically the equivalent of what I'd do in C with a `struct`.  And because things aren't pass by pointer/reference, the counter variable is hacked up to be an array.  Perhaps I should just make an object to hold all of this and call it a day?

**Comment by grondo** [comment]
> +def list_id_cb(future, arg):
> +    (flux_handle, rpc_handle, jobid, count, args, jobs) = arg

AFAIK, the `future` passed to `list_id_cb` here *is* an rpc handle, so the `rpc_handle` is redundant.

> And because things aren't pass by pointer/reference, the counter variable is hacked up to be an array.  Perhaps I should just make an object to hold all of this and call it a day?

I think it would be slightly more natural to pass a table as `arg`, or have an internal class as you suggest (some might think that would be overkill though).

**Comment by chu11** [comment]
rebased & re-pushed, a few minor cleanups and added tests.  Removing WIP as its approaching something mergable.

**Comment by chu11** [comment]
Small aside, I decided to have output like the following when the user inputs job ids that are invalid.  Output an error message indicating a bad ID, but output all IDs that are valid.


```
>flux jobs 433422598144 01234
JobID 1234 unknown
             JOBID USER     NAME       STATE    NTASKS NNODES  RUNTIME RANKS
      433422598144 achu     hostname   INACTIVE      1      1   0.076s 0
```

**Comment by grondo** [comment]
Looks like other job selection options are silently ignored if JOBIDs are listed on the command line? This matches the description in the manpage, so no surprise there, but I wonder if a warning should be emitted if any of those options are used, but explicit JOBIDs are listed?

File: src/cmd/flux-jobs.py [hunk 1]
```diff
@@ -190,6 +191,44 @@ def fetch_jobs_stdin():
     return jobs
 
 
+def list_id_cb(future, arg):
+    (flux_handle, jobid, count, args, jobs) = arg
```
[the remark this answers is not in the document]
Comment (chu11) [reply src/cmd/flux-jobs.py:195]: > However, I'm not sure if the way you're doing it here is more Pythonic (@SteVwonder did give you thumbs up on your original proposal..) So feel free to ignore me.

yeah, a class seemed overkill when I thought about it, but now that I see there are simpler ways to initialize a dict than what I knew, that might work out more cleanly.

**Comment by chu11** [comment]
> Looks like other job selection options are silently ignored if JOBIDs are listed on the command line? This matches the description in the manpage, so no surprise there, but

Yeah, its mostly due to the fact we use the `job-info.list-id` service.  If we were to honor the other command line options, we'd just be doing a glorified grep for the user.

As I mentioned above, we could make the `job-info.list-id` service take a list of IDs.  Then we could add filtering options just like with the other list services.

> I wonder if a warning should be emitted if any of those options are used, but explicit JOBIDs are listed?

Seems like a good idea.

**Comment by chu11** [comment]
re-pushed with a cleanup that does the following for the callback, I agree with @grondo it looks cleaner.

```
    cbargs = {
        "flux_handle" : flux_handle,
        "jobs" : [],
        "count" : 0,
        "total" : len(args.jobids)
    }
    for jobid in args.jobids:
        rpc_handle = flux.job.job_list_id(flux_handle, jobid, list(attrs))
        rpc_handle.then(list_id_cb, arg=(cbargs, jobid))
```

One thing I forgot, is that I couldn't entirely remove the tuple.  There are "global" args for every callback and "local" ones for each callback (most notably jobid).  But I think it's a net win readability wise.

> I wonder if a warning should be emitted if any of those options are used, but explicit JOBIDs are listed?

I looked into this and realized using Python's parseargs, I don't know if an option was set by the user or if the defaults are set.  I think there's trickery that can be done to do this, but not sure if it's worth it.  Unless there's a clean trick I don't know :P

**Comment by grondo** [comment]
> I looked into this and realized using Python's parseargs, I don't know if an option was set by the user or if the defaults are set. I think there's trickery that can be done to do this, but not sure if it's worth it. Unless there's a clean trick I don't know :P

Looks like you could add a new [`Action`](https://docs.python.org/3.6/library/argparse.html#argparse.Action) class for relevant options, extending the `store` and `store_true` actions to *also* set some attribute in the returned namespace (e.g. `filter_active`), to indicate one of a set of options was used.

However, I agree it would be a bit of work for not much present gain.

**Comment by chu11** [comment]
> Looks like you could add a new Action class for relevant options, extending the store and store_true actions to also set some attribute in the returned namespace (e.g. filter_active), to indicate one of a set of options was used.

Yeah, that was the first thing I thought of, overriding `action`, but it seemed like a lot.

**Comment by chu11** [comment]
hmmm travis builder failed

```
FAIL: t0020-terminus.t 31 - flux-terminus: copy stdin

          t0020-terminus.t:  FAIL: N=34  PASS=33  FAIL=1 SKIP=0 XPASS=0 XFAIL=0

ERROR: t0020-terminus.t - exited with status 1
```

```
expecting success: 
	$runpty -o log.pipe-stdin flux terminus start &
	pid=$! &&
	$waitfile -t 20 -v -p \"o\" log.pipe-stdin &&
	printf "echo hello\r" | flux terminus attach -p 0 &&
	$waitfile -t 20 -v -p hello log.pipe-stdin &&
	printf "exit\r" | flux terminus attach -p 0 &&
	$waitfile -t 20 -v -p detached log.pipe-stdin &&
	wait $pid

waitfile: log.pipe-stdin: 0.000s: No file yet
waitfile: log.pipe-stdin: 0.251s: wakeup
waitfile: log.pipe-stdin: 0.251s: File appeared with size 322
waitfile: log.pipe-stdin: 0.251s: log.pipe-stdin seek set to 0
waitfile: log.pipe-stdin: 0.251s: Got match (1/1)
waitfile: log.pipe-stdin: 0.251s: Got match. Calling self:on_match()
waitfile: log.pipe-stdin: 0.251s: Exiting
{"version": 2, "width": 80, "height": 25, "timestamp": 1587600020}
[0.03823256492614746, "o", "\u001b[2J\u001b[;H"]
waitfile: log.pipe-stdin: 0.000s: File appeared with size 399
waitfile: log.pipe-stdin: 0.000s: log.pipe-stdin seek set to 0
waitfile: log.pipe-stdin: 0.000s: Got match (1/1)
waitfile: log.pipe-stdin: 0.000s: Got match. Calling self:on_match()
waitfile: log.pipe-stdin: 0.000s: Exiting
{"version": 2, "width": 80, "height": 25, "timestamp": 1587600020}
[0.03823256492614746, "o", "\u001b[2J\u001b[;H"]
[0.042634010314941406, "o", "To run a command as administrator (user \"root\"), use \"sudo <command>\".\r\nSee \"man sudo_root\" for details.\r\n\r\n"]
[0.04330277442932129, "o", "travis@6e01950176fc:~$ "]
[0.20807552337646484, "o", "echo hello\r\nhello\r\ntravis@6e01950176fc:~$ "]
waitfile: log.pipe-stdin: 0.000s: File appeared with size 476
waitfile: log.pipe-stdin: 0.000s: log.pipe-stdin seek set to 0
waitfile: log.pipe-stdin: 0.000s: leaving checklines() position=476
waitfile: log.pipe-stdin: 20.018s: Timeout after 20s
-rw-r--r-- 1 travis travis 476 Apr 23 00:00 log.pipe-stdin
{"version": 2, "width": 80, "height": 25, "timestamp": 1587600020}
[0.03823256492614746, "o", "\u001b[2J\u001b[;H"]
[0.042634010314941406, "o", "To run a command as administrator (user \"root\"), use \"sudo <command>\".\r\nSee \"man sudo_root\" for details.\r\n\r\n"]
[0.04330277442932129, "o", "travis@6e01950176fc:~$ "]
[0.20807552337646484, "o", "echo hello\r\nhello\r\ntravis@6e01950176fc:~$ "]
[0.2577371597290039, "o", "exit\r\nexit"]
[0.25838804244995117, "o", "\r\n"]
not ok 31 - flux-terminus: copy stdin

expecting success: 
	flux terminus kill-server
	$runpty -o log.disconnect flux terminus start &
	pid=$! &&
	$waitfile -t 20 -v -p \"o\" log.disconnect &&
	kill -TERM $pid &&
	test_expect_code 143 wait $pid &&
	flux terminus list >disconnect.list &&
	test_debug "cat disconnect.list" &&
	grep -q "0 clients" disconnect.list &&
	flux terminus kill 0
```

@grondo want to take a look before I restart it?

**Comment by grondo** [comment]
I think you pasted enough above to try to figure out what went wrong. Go ahead and restart. Thanks.

**Comment by grondo** [comment]
> but it seemed like a lot.

For my own edification, I tried it out:

```diff
diff --git a/src/cmd/flux-jobs.py b/src/cmd/flux-jobs.py
index 661df5a59..d4f1e679e 100755
--- a/src/cmd/flux-jobs.py
+++ b/src/cmd/flux-jobs.py
@@ -319,19 +319,53 @@ def fetch_jobs(args, fields):
     return [JobInfo(job) for job in lst]
 
 
+class FilterAction(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        setattr(namespace, self.dest, values)
+        setattr(namespace, "filtered", True)
+
+
+class FilterTrueAction(argparse.Action):
+    def __init__(
+        self,
+        option_strings,
+        dest,
+        const=True,
+        default=False,
+        required=False,
+        help=None,
+        metavar=None,
+    ):
+        super(FilterTrueAction, self).__init__(
+            option_strings=option_strings,
+            dest=dest,
+            nargs=0,
+            const=const,
+            default=default,
+            help=help,
+        )
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        setattr(namespace, self.dest, self.const)
+        setattr(namespace, "filtered", True)
+
+
 def parse_args():
     parser = argparse.ArgumentParser(
         prog="flux-jobs", formatter_class=flux.util.help_formatter()
     )
     # -a equivalent to -s "pending,running,inactive" and -u set to userid
     parser.add_argument(
-        "-a", action="store_true", help="List all jobs for current user"
+        "-a", action=FilterTrueAction, help="List all jobs for current user"
     )
     # -A equivalent to -s "pending,running,inactive" and -u set to "all"
-    parser.add_argument("-A", action="store_true", help="List all jobs for all users")
+    parser.add_argument(
+        "-A", action=FilterTrueAction, help="List all jobs for all users"
+    )
     parser.add_argument(
         "-c",
         "--count",
+        action=FilterAction,
         type=int,
         metavar="N",
         default=1000,
@@ -340,6 +374,7 @@ def parse_args():
     parser.add_argument(
         "-s",
         "--states",
+        action=FilterAction,
         type=str,
         metavar="STATES",
         default="pending,running",
@@ -354,6 +389,7 @@ def parse_args():
     parser.add_argument(
         "-u",
         "--user",
+        action=FilterAction,
         type=str,
         metavar="[USERNAME|UID]",
         default=str(os.getuid()),
@@ -376,6 +412,7 @@ def parse_args():
     )
     # Hidden '--from-stdin' option for testing only.
     parser.add_argument("--from-stdin", action="store_true", help=argparse.SUPPRESS)
+    parser.set_defaults(filtered=False)
     return parser.parse_args()
 
 
@@ -479,6 +516,10 @@ class JobsOutputFormat(flux.util.OutputFormat):
 def main():
     args = parse_args()
 
+    if args.jobids and args.filtered:
+        LOGGER.error("Do not use filter options (-a,-A,-c,-u,-s) with jobid list")
+        sys.exit(1)
+
     if args.format:
         fmt = args.format
     else:

```


```console
grondo@asp:~/git/flux-core.git$ flux jobs -a 1234
flux-jobs: ERROR: Do not use filter options (-a,-A,-c,-u,-s) with jobid list
```

**Comment by grondo** [comment]
Rebasing since it seems like the GH actions aren't reporting.

@mergifyio rebase

**Comment by chu11** [comment]
> For my own edification, I tried it out:

@grondo nice!  Thanks for trying it out.  

Doesn't seem too bad, of course we're somewhat fortunate that we only have two actions to do.  It would probably get uglier if there were more.  I think the patch can do on top if you think it does.  Then I'll just add a test too.

**Comment by grondo** [comment]
> Doesn't seem too bad, of course we're somewhat fortunate that we only have two actions to do. It would probably get uglier if there were more. I think the patch can do on top if you think it does. Then I'll just add a test too.

Yeah, though other actions might be able to just subclass one of the other two.

I'll leave it up to you if you want to include this functionality or not.

**Comment by chu11** [comment]
> I'll leave it up to you if you want to include this functionality or not.

Ok I'll add it.

I just realized why this is simpler than I imagined.  I was going to output a warning per option, but your approach uses just a flag to indicate if any of the filtering options was set.  Which is probably more than fine.

**Comment by chu11** [comment]
re-pushed w/ @grondo's commit and added a simple test.  I did tweak the output so its a WARNING instead of an error & exit.

pylint seems to hate the fact the Action class has a variable called `help`, so had to add a pylint disable in there.

**Comment by SteVwonder** [comment]
> I think it would be slightly more natural to pass a table as arg, or have an internal class as you suggest (some might think that would be overkill though).

Sorry to be late to comment.  For little one-off things like this, Python has a namedtuple: https://docs.python.org/3/library/collections.html#collections.namedtuple

**Comment by chu11** [comment]
> Sorry to be late to comment. For little one-off things like this, Python has a namedtuple: https://docs.python.org/3/library/collections.html#collections.namedtuple

Yeah, I found that online somewhere.  But in this particular case I need a data structure that is mutable, and `namedtuple` appears to be immutable :-(

**Comment by grondo** [comment]
Looked like asan hung, so I restarted it.

**Comment by chu11** [comment]
Thanks!  And thanks for the actions filtering support.

**Pull request merged**